src/stream/src/common/builder.rs

```diff
@@ -200,3 +144,120 @@ impl StreamChunkBuilder {
         ))
     }
 }
+
+type IndexMappings = Vec<(usize, usize)>;
+
+/// Build stream chunks with fixed chunk size from joined two sides of rows.
+pub struct JoinStreamChunkBuilder {
+    builder: StreamChunkBuilder,
+
+    /// The column index mapping from update side to output.
+    update_to_output: IndexMappings,
+
+    /// The column index mapping from matched side to output.
+    matched_to_output: IndexMappings,
+}
+
+impl JoinStreamChunkBuilder {
+    pub fn new(
+        chunk_size: usize,
+        data_types: Vec<DataType>,
+        update_to_output: IndexMappings,
+        matched_to_output: IndexMappings,
+    ) -> Self {
```

Comment:
Could you give an example in the doc to clarify these index mappings?

Reply:
And what about replacing the term `update` and `matched` with simply `left` and `right`? Because from builder's view there is only left and right, and I can't tell which one is on which side if I don't read the code.

Reply:
I guess it's not simply `left` and `right`. In HashJoin, the two side can be switched, the `update` side is the side which update rows come, it can be left or right. And, in the `append_row_update/matched` method, the rows is not necessary arranged as `update` on the left and `matched` on the right, the mapping is determined by `update_to_output` and `matched_to_output`.

Your next question can be answered together with this one. The mappings are not simply `output_indices`, so seems can't be reduced to `Vec<usize>`. (Actually I tried to store `output_indices` instead of `update_to_output` and `matched_to_output` in `JoinStreamChunkBuilder`, and failed to (elegantly) adopt it to HashJoin executor😇)

Reply:
Took me some time to understand the difference between `left` `right` and `update` `matched`. I've tried my best to use `left` `right` and `update` `matched` correctly in the code(

Reply:
Okay I can understand a little bit. Then it seems more necessary to have an example to explain the possible situations of `updated` and `matched`. 😇